File: leetcode/src/java/SingleNumber.java
import java.util.HashSet;
import java.util.Set;

public class SingleNumber {
    public static void main(String[] args) {
        SingleNumber instance = new SingleNumber();
        System.out.println(instance.singleNumber(new int[]{4, 1, 2, 1, 2}));
        System.out.println(instance.singleNumber(new int[]{1}));
    }
    public int singleNumber(int[] nums) {
        Set<Integer> set = new HashSet<>();
        for (int num : nums) {
            if (set.contains(num)) {
                set.remove(num);
            } else {
                set.add(num);
            }
        }
        return set.toArray(new Integer[]{})[0];
    }
}

File: leetcode/src/python/single_number.py
from typing import List

class SingleNumber:
    def singleNumber(self, nums: List[int]) -> int:
        s = set()
        for num in nums:
            if num in s:
                s.remove(num)
            else:
                s.add(num)
        return s.pop()

if __name__ == '__main__':
    instance = SingleNumber()
    nums = [4, 1, 2, 1, 2]
    print(instance.singleNumber(nums))